keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,880 @@
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The design proposes adding a new KubeletConfig field `ShutdownGracePeriod` used
+to specify total time period kubelet should delay shutdown by and thus total time
+allocated to the graceful termination process.
+
+In addition to `ShutdownGracePeriod`, another KubeletConfig field will be added
+`ShutdownGracePeriodCriticalPods`. During the shutdown, the
+`ShutdownGracePeriod-ShutdownGracePeriodCriticalPods` duration will be grace
+period for non critical system pods like user workloads, while the remaining
+time of `ShutdownGracePeriodCriticalPods` will be the grace period for critical
+pods like node logging daemonsets.
+
+```
+type KubeletConfiguration struct {
+    ...
+    ShutdownGracePeriod metav1.Duration
+    ShutdownGracePeriodCriticalPods metav1.Duration
+}
+```
+
+Communication with systemd over dbus for (creating inhibitor lock, receiving
+`PrepareForShutdown` callback, etc), will make use of the
+`github.com/godbus/dbus/v5` package which is already included in
+[`vendor/`](https://github.com/kubernetes/kubernetes/tree/release-1.19/vendor/github.com/godbus/dbus/v5).
+
+Termination of pods will make use of the existing
+[killPod](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/kubelet/pod_workers.go#L292)function
+from the `kubelet` package and specify the appropriate `gracePeriodOverride` as
+necessary.
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+*   Unit tests for kubelet of handling shutdown event
+*   New E2E tests to validate node graceful shutdown (note limitation that K8S
+    E2E tests currently only run on GCE).
+    *   Shutdown grace period unspecified, feature is not active
+    *   Pod’s ExecStop and SIGTERM handlers are given gracePeriodSeconds for
+        case when gracePeriodSeconds <= kubeletConfig.ShutdownGracePeriod
+    *   Pod’s ExecStop and SIGTERM handlers are given
+        kubeletConfig.ShutdownGracePeriod for case when gracePeriodSeconds >
+        kubeletConfig.ShutdownGracePeriod
+
+### Graduation Criteria
+
+<!--
```

Comment:
Need to add something here for alpha in 1.20

Reply:
Added.